[docs] update web publishing docs

marklawlor opened the pull request:
# Why

Update the web publishing docs with more information around Expo Router's output modes.

We are also trying to shifted the tone away from "Metro vs webpack" and treat Expo Router as the best way to publish. So I've tried to focus more on Expo Router's output modes.

@amandeepmittal I don't like my layout, I think I'm using too many components back-to-back and it looks very blocky. Would love some feedback.

@kitten We should also work together to update the platforms section

marklawlor committed:
(docs): update web publishing docs

marklawlor commented on docs/pages/distribution/publishing-websites.mdx:
```diff
@@ -3,12 +3,56 @@ title: Publish websites
+
+> If you are using Expo Web without Expo Router, only `single` is supported
+
+| Output             | Expo Router | API Routes  |  Description                                                                                                                   |
```
I would like advise on this table. I'm trying to explain the outputs, while also showing the API routes only works with 1 output, while also explaining that these are Expo Router only, and Metro/Webpack don't have these features.

Its got a lot going on

amandeepmittal replied:
I think the table looks fine. To add more context of what you are trying to explain, let's add it above this table.

marklawlor commented on docs/pages/distribution/publishing-websites.mdx:
```diff
@@ -55,20 +119,28 @@ Use [Serve CLI](https://www.npmjs.com/package/serve) to quickly test locally how
 
+<Tab label="server">
+
+There is no prebuilt solution. You will need to create your own server to serve the static files and API routes.
```
This statement feels bad. How can I improve the language here? I tried to get `npx serve` working with this output but I gave up.

@kitten Got any advise?

EvanBacon commented on docs/pages/distribution/publishing-websites.mdx:
```diff
@@ -3,12 +3,56 @@ title: Publish websites
 
+## Output targets
+
+Expo supports three output targets for web apps. The output target can be configured in the **app.json** file:
```
```suggestion
Expo Router supports three output targets for web apps. The `web.output` target can be configured in the **app.json** file:
```

EvanBacon commented on docs/pages/distribution/publishing-websites.mdx:
````diff
@@ -3,12 +3,56 @@ title: Publish websites
+}
+```
+
+> If you are using Expo Web without Expo Router, only `single` is supported
````
```suggestion
> If you are using legacy Expo Webpack (deprecated), only `single` is supported.
```

marklawlor committed:
Update docs/pages/distribution/publishing-websites.mdx

Co-authored-by: Evan Bacon <bacon@expo.io>

marklawlor committed:
Update docs/pages/distribution/publishing-websites.mdx

Co-authored-by: Evan Bacon <bacon@expo.io>

amandeepmittal commented on docs/pages/distribution/publishing-websites.mdx:
```diff
@@ -3,12 +3,56 @@ title: Publish websites
 
+## Output targets
+
+Expo Router supports three output targets for web apps. The `web.output` target can be configured in the **app.json** file:
```
```suggestion
The [`web.output`](/versions/latest/config/app/#output) target can be configured in the [app config](/workflow/configuration/) to set the export method for the web app:
```

Let's add link to the `web.output` config definition and also let's refer app config instead of `app.json`. It's what we use for consistency. Let's also briefly explain what it does.

amandeepmittal commented on docs/pages/distribution/publishing-websites.mdx:
````diff
@@ -3,12 +3,56 @@ title: Publish websites
+```
+
+> If you are using legacy Expo Webpack (deprecated), only `single` is supported.
+
````
```suggestion

Expo Router supports three output targets for web apps.

```

amandeepmittal commented on docs/pages/distribution/publishing-websites.mdx:
```diff
@@ -3,12 +3,56 @@ title: Publish websites
+
+| Output             | Expo Router | API Routes  |  Description                                                                                                                   |
+| ------------------ | ----------- | ----------- |  ----------------------------------------------------------------------------------------------------------------------------- |
+| `single` (default) | <YesIcon /> | <NoIcon />  |  Outputs a Single Page Application (SPA), with a single index.html in the output folder, and has no statically indexable HTML. |
```
```suggestion
| `single` (default) | <YesIcon /> | <NoIcon />  |  Outputs a Single Page Application (SPA) with a single **index.html** in the output folder and has no statically indexable HTML. |
```

Let's remove the commas and highlight the file name in bold per our writing style guide.

amandeepmittal commented on docs/pages/distribution/publishing-websites.mdx:
```diff
@@ -3,12 +3,56 @@ title: Publish websites
+| Output             | Expo Router | API Routes  |  Description                                                                                                                   |
+| ------------------ | ----------- | ----------- |  ----------------------------------------------------------------------------------------------------------------------------- |
+| `single` (default) | <YesIcon /> | <NoIcon />  |  Outputs a Single Page Application (SPA), with a single index.html in the output folder, and has no statically indexable HTML. |
+| `server`           | <YesIcon /> | <YesIcon /> |  Creates a client & server folder. Client files are outputed as seperate HTML files. API routes as seperate JavaScript files   |
```
```suggestion
| `server`           | <YesIcon /> | <YesIcon /> |  Creates **client** and **server** directories. Client files are output as separate HTML files. API routes as separate JavaScript files for hosting with a custom Node.js server.   |
```

Added a bit more context for the API routes part of this description based on https://docs.expo.dev/versions/latest/config/app/#output. It wasn't clear what would one do with API routes here or why they should know about it.

amandeepmittal commented on docs/pages/distribution/publishing-websites.mdx:
```diff
@@ -3,12 +3,56 @@ title: Publish websites
+| ------------------ | ----------- | ----------- |  ----------------------------------------------------------------------------------------------------------------------------- |
+| `single` (default) | <YesIcon /> | <NoIcon />  |  Outputs a Single Page Application (SPA), with a single index.html in the output folder, and has no statically indexable HTML. |
+| `server`           | <YesIcon /> | <YesIcon /> |  Creates a client & server folder. Client files are outputed as seperate HTML files. API routes as seperate JavaScript files   |
+| `static`           | <YesIcon /> | <NoIcon />  |  Outputs seperate HTML files for every route in the app/ directory                                                             |
```
```suggestion
| `static`           | <YesIcon /> | <NoIcon />  |  Outputs separate HTML files for every route in the **app** directory.                                                             |
```

amandeepmittal commented on docs/pages/distribution/publishing-websites.mdx:
````diff
@@ -3,12 +3,56 @@ title: Publish websites
+}
+```
+
+> If you are using legacy Expo Webpack (deprecated), only `single` is supported.
````
```suggestion
```

We can combine the info in this callout with webpack info later within this section. Having two contexts for the same information above and below the table didn't make sense to me.

amandeepmittal commented on docs/pages/distribution/publishing-websites.mdx:
```diff
@@ -3,12 +3,56 @@ title: Publish websites
+| `server`           | <YesIcon /> | <YesIcon /> |  Creates a client & server folder. Client files are outputed as seperate HTML files. API routes as seperate JavaScript files   |
+| `static`           | <YesIcon /> | <NoIcon />  |  Outputs seperate HTML files for every route in the app/ directory                                                             |
+
+> **warning** In SDK 50+, publishing with `webpack` is deprecated
```
```suggestion
> **warning** In SDK 50 and above, publishing with `webpack` is deprecated. 
```

amandeepmittal commented on docs/pages/distribution/publishing-websites.mdx:
```diff
@@ -3,12 +3,56 @@ title: Publish websites
+
+> **warning** In SDK 50+, publishing with `webpack` is deprecated
+
+Expo only supports the single output format when using `webpack`.
```
```suggestion
Expo only supports the `single` output format using `webpack`.
```

amandeepmittal commented on docs/pages/distribution/publishing-websites.mdx:
```diff
@@ -3,12 +3,56 @@ title: Publish websites
+
+| Bundler   | Expo Router | API Routes | Description                                                                             |
+| --------- | ----------- | ---------- | --------------------------------------------------------------------------------------- |
+| `webpack` | <NoIcon />  | <NoIcon /> | Outputs a Single Page Application (SPA), with a single index.html in the output folder. |
```
```suggestion
| `webpack` | <NoIcon />  | <NoIcon /> | Outputs a Single Page Application (SPA) with a single **index.html** in the output folder. |
```

amandeepmittal commented on docs/pages/distribution/publishing-websites.mdx:
```diff
@@ -55,20 +119,28 @@ Use [Serve CLI](https://www.npmjs.com/package/serve) to quickly test locally how
+<Terminal cmd={['$ npx serve dist --single']} />
 
-<Terminal cmd={['$ npx serve dist']} />
+Open [`http://localhost:5000`](http://localhost:5000) to see your project in action. This method is **HTTP only**, so permissions, camera, location, and many other secure features won't work.
```
```suggestion
Open [`http://localhost:5000`](http://localhost:5000) to see your project in action. This method is **HTTP only**, so permissions, camera, location, and many other secure features will not work.
```

Let's avoid the contraction here.

amandeepmittal commented on docs/pages/distribution/publishing-websites.mdx:
```diff
@@ -55,20 +119,28 @@ Use [Serve CLI](https://www.npmjs.com/package/serve) to quickly test locally how
+
+<Terminal cmd={['$ npx serve dist']} />
 
 Open [`http://localhost:5000`](http://localhost:5000) to see your project in action. This method is **HTTP only**, so permissions, camera, location, and many other secure features won't work.
```
```suggestion
Open [`http://localhost:5000`](http://localhost:5000) to see your project in action. This method is **HTTP only**, so permissions, camera, location, and many other secure features will work.
```

I just realized this is outside the scope of this PR so feel free to ignore this and the other similar suggestion.

marklawlor committed:
Update docs/pages/distribution/publishing-websites.mdx

Co-authored-by: Aman Mittal <amandeepmittal@live.com>

marklawlor committed:
Update docs/pages/distribution/publishing-websites.mdx

Co-authored-by: Aman Mittal <amandeepmittal@live.com>

marklawlor committed:
Update docs/pages/distribution/publishing-websites.mdx

Co-authored-by: Aman Mittal <amandeepmittal@live.com>

marklawlor committed:
Update docs/pages/distribution/publishing-websites.mdx

Co-authored-by: Aman Mittal <amandeepmittal@live.com>

marklawlor committed:
Update docs/pages/distribution/publishing-websites.mdx

Co-authored-by: Aman Mittal <amandeepmittal@live.com>

marklawlor committed:
Update docs/pages/distribution/publishing-websites.mdx

Co-authored-by: Aman Mittal <amandeepmittal@live.com>

marklawlor committed:
Update docs/pages/distribution/publishing-websites.mdx

Co-authored-by: Aman Mittal <amandeepmittal@live.com>

marklawlor committed:
Update docs/pages/distribution/publishing-websites.mdx

Co-authored-by: Aman Mittal <amandeepmittal@live.com>

marklawlor committed:
Update docs/pages/distribution/publishing-websites.mdx

Co-authored-by: Aman Mittal <amandeepmittal@live.com>

marklawlor committed:
Update docs/pages/distribution/publishing-websites.mdx

Co-authored-by: Aman Mittal <amandeepmittal@live.com>

marklawlor committed:
Update docs/pages/distribution/publishing-websites.mdx

Co-authored-by: Aman Mittal <amandeepmittal@live.com>

marklawlor committed:
split publishing websites into two seperate pages

marklawlor commented:
@amandeepmittal Can you please re-review. I've added your changes and also split the "publishing websites" page into two (for each blunder)

I've kept the same slug for the Metro page so no links are broken

amandeepmittal commented on docs/pages/distribution/publishing-websites-webpack.mdx:
```diff
@@ -0,0 +1,337 @@
+---
+title: Publish websites with Webpack
+description: Different ways to publish the Expo web app with third-party services.
```
```suggestion
description: Different ways to publish the Expo web app with third-party services with webpack.
```

amandeepmittal commented on docs/pages/distribution/publishing-websites-webpack.mdx:
```diff
@@ -0,0 +1,337 @@
+
+## Create a build
+
+Creating a build of the project is the first step to publishing a web app. Whether you want to serve it locally or host it on a third-party service, you'll need to export all JavaScript and assets of a project. This is known as a static bundle. It can be exported by running the following command:
```
```suggestion
Creating a build of the project is the first step to publishing a web app. Whether you want to serve it locally or host it on a third-party service, you have to export all JavaScript and assets of a project. This is known as a static bundle. It can be exported by running the following command:
```

- Let's avoid contractions.

amandeepmittal commented on docs/pages/distribution/publishing-websites-webpack.mdx:
```diff
@@ -0,0 +1,337 @@
+
+The resulting project files are in the **web-build** directory. Any files inside the **web** directory are also copied to the **web-build** directory.
+
+If you make changes to your project, you'll need to rebuild it for production. Don't edit the **web-build** directory directly.
```
```suggestion
If you make changes to your project, rebuild it for production. Do not edit the **web-build** directory directly.
```

amandeepmittal commented on docs/pages/distribution/publishing-websites-webpack.mdx:
````diff
@@ -0,0 +1,337 @@
+
+If you make changes to your project, you'll need to rebuild it for production. Don't edit the **web-build** directory directly.
+
+> **Tip:** If you want to serve your site in a subfolder, add its path to your **package.json** as shown below:
+>
+> ```json package.json
+> {
+>   "homepage": "/path/to/subfolder"
+> }
+> ```
````
````suggestion
> **Tip:** If you want to serve your site in a subfolder, add its path to your **package.json** as shown below:
>
> ```json package.json
> {
>   "homepage": "/path/to/subfolder"
> }
> ```
````

Let's remove the callout here. For consistency, I'd like us to avoid adding code snippets (other than inline ones)  in callouts. It also has broken styles.

![CleanShot 2023-12-19 at 02 06 23](https://github.com/expo/expo/assets/10234615/6494ae21-b62d-4763-aa9c-1c4a7fc08fa7)


Let's add this as a part of the main content. And since this is a "Tip", let's use **info** callout type here.

amandeepmittal commented on docs/pages/distribution/publishing-websites-webpack.mdx:
````diff
@@ -0,0 +1,337 @@
+
+Navigate inside the **web-build** directory and run the following command to create **\_redirects** file with follwing rule:
+
+```sh web/_redirects
````
````suggestion
```bash web/_redirects
````

amandeepmittal commented on docs/pages/distribution/publishing-websites-webpack.mdx:
```diff
@@ -0,0 +1,337 @@
+
+<Terminal cmd={['$ netlify deploy --dir web-build']} />
+
+You'll see a URL that you can use to view your project online.
```
```suggestion
You will see a URL that you can use to view your project online.
```

amandeepmittal commented on docs/pages/distribution/publishing-websites-webpack.mdx:
```diff
@@ -0,0 +1,337 @@
+
+</Step>
+<Step label="2">
+  Configure redirects for single-page applications.
+  Create a **vercel.json** file at the root of your app and add the following:
```
```suggestion
  Configure redirects for single-page applications. Create a **vercel.json** file at the root of your app and add the following:
```

amandeepmittal commented on docs/pages/distribution/publishing-websites-webpack.mdx:
```diff
@@ -0,0 +1,337 @@
+
+<Terminal cmd={['$ vercel']} />
+
+You'll now see a URL that you can use to view your project online. Paste that URL into your browser when the build is complete, and you'll see your deployed app.
```
```suggestion
You will now see a URL that you can use to view your project online. Paste that URL into your browser when the build is complete, and see your deployed app.
```

amandeepmittal commented on docs/pages/distribution/publishing-websites-webpack.mdx:
```diff
@@ -0,0 +1,337 @@
+
+### AWS Amplify Console
+
+The [AWS Amplify Console](https://console.amplify.aws) provides a Git-based workflow for continuously deploying and hosting full-stack serverless web apps. Amplify deploys your PWA from a repository instead of from your computer. In this guide, we'll use a GitHub repository. Before starting, [create a new repo on GitHub](https://github.com/new).
```
```suggestion
The [AWS Amplify Console](https://console.amplify.aws) provides a Git-based workflow for continuously deploying and hosting full-stack serverless web apps. Amplify deploys your PWA from a repository instead of your computer, so you must use a GitHub repository. Before starting, [create a new repo on GitHub](https://github.com/new).
```

amandeepmittal commented on docs/pages/distribution/publishing-websites-webpack.mdx:
```diff
@@ -0,0 +1,337 @@
+
+<Step label="2">
+
+Push your local Expo project to a GitHub repository. If you haven't pushed to GitHub yet, follow [GitHub's guide to add an existing project to GitHub](https://docs.github.com/en/get-started/importing-your-projects-to-github/importing-source-code-to-github/adding-locally-hosted-code-to-github).
```
```suggestion
Push your local Expo project to a GitHub repository. If you have not pushed to GitHub yet, follow [GitHub's guide to add an existing project to GitHub](https://docs.github.com/en/get-started/importing-your-projects-to-github/importing-source-code-to-github/adding-locally-hosted-code-to-github).
```

amandeepmittal commented on docs/pages/distribution/publishing-websites.mdx:
```diff
@@ -55,27 +82,25 @@ Use [Serve CLI](https://www.npmjs.com/package/serve) to quickly test locally how
 
-<Tab label="webpack">
+<Tab label="static">
+
+<Terminal cmd={['$ npx serve dist']} />
 
-Run the following command to serve the single-page application:
+Open [`http://localhost:5000`](http://localhost:5000) to see your project in action. This method is **HTTP only**, so permissions, camera, location, and many other secure features will work.
+
+</Tab>
```
Let's move this after the server since it seems to be the pattern for this whole document (and we should follow it even though it might give our users a tiny but uniformed experience).

marklawlor commented:
Feedback: Add callout at top of Metro page to link to webpack, as we are highjacking the url

marklawlor committed:
Update docs/pages/distribution/publishing-websites-webpack.mdx

Co-authored-by: Aman Mittal <amandeepmittal@live.com>

marklawlor committed:
Update docs/pages/distribution/publishing-websites-webpack.mdx

Co-authored-by: Aman Mittal <amandeepmittal@live.com>

marklawlor committed:
Update docs/pages/distribution/publishing-websites-webpack.mdx

Co-authored-by: Aman Mittal <amandeepmittal@live.com>

marklawlor committed:
Update docs/pages/distribution/publishing-websites-webpack.mdx

Co-authored-by: Aman Mittal <amandeepmittal@live.com>

marklawlor committed:
Update docs/pages/distribution/publishing-websites-webpack.mdx

Co-authored-by: Aman Mittal <amandeepmittal@live.com>

marklawlor committed:
Update docs/pages/distribution/publishing-websites-webpack.mdx

Co-authored-by: Aman Mittal <amandeepmittal@live.com>

marklawlor committed:
Update docs/pages/distribution/publishing-websites-webpack.mdx

Co-authored-by: Aman Mittal <amandeepmittal@live.com>

marklawlor committed:
Update docs/pages/distribution/publishing-websites-webpack.mdx

Co-authored-by: Aman Mittal <amandeepmittal@live.com>

marklawlor committed:
Update docs/pages/distribution/publishing-websites-webpack.mdx

Co-authored-by: Aman Mittal <amandeepmittal@live.com>

marklawlor committed:
pr feedback

marklawlor commented:
@amandeepmittal Can you take one final look at this before we merge this for the SDK 50 release?

amandeepmittal commented on docs/pages/distribution/publishing-websites-webpack.mdx:
```diff
@@ -0,0 +1,336 @@
+import { Terminal } from '~/ui/components/Snippet';
+import { NoIcon } from '~/ui/components/DocIcons';
+
+> **warning** In SDK 50 and above, publishing with `webpack` is deprecated.
```
```suggestion
> **warning** **Deprecation**: In SDK 50 and above, publishing with `webpack` is deprecated.
```

amandeepmittal commented on docs/pages/distribution/publishing-websites-webpack.mdx:
```diff
@@ -0,0 +1,336 @@
+| --------- | ----------- | ---------- | ------------------------------------------------------------------------------------------ |
+| `webpack` | <NoIcon />  | <NoIcon /> | Outputs a Single Page Application (SPA) with a single **index.html** in the output folder. |
+
+## Create a build
```
```suggestion
## Create a web build
```

amandeepmittal commented on docs/pages/distribution/publishing-websites-webpack.mdx:
```diff
@@ -0,0 +1,336 @@
+
+## Create a build
+
+Creating a build of the project is the first step to publishing a web app. Whether you want to serve it locally or host it on a third-party service, you have to export all JavaScript and assets of a project. This is known as a static bundle. It can be exported by running the following command:
```
```suggestion
Creating a web build of the project is the first step to publishing a web app. Whether you want to serve it locally or host it on a third-party service, you have to export all JavaScript and assets of a project. This is known as a static bundle. It can be exported by running the following command:
```

amandeepmittal reviewed (approved):
Lgtm @marklawlor 👍 

Thank you for updating and improving these docs.

Left a few minor suggestions to improve a section title and warning callout.

marklawlor committed:
Update docs/pages/distribution/publishing-websites-webpack.mdx

Co-authored-by: Aman Mittal <amandeepmittal@live.com>

marklawlor committed:
Update docs/pages/distribution/publishing-websites-webpack.mdx

Co-authored-by: Aman Mittal <amandeepmittal@live.com>

marklawlor committed:
Update docs/pages/distribution/publishing-websites-webpack.mdx

Co-authored-by: Aman Mittal <amandeepmittal@live.com>
